src/lua/uri.lua

```diff
@@ -280,26 +281,12 @@ local function escape(buf, opts)
     options.unreserved = options.unreserved or RFC3986.unreserved
     options.plus = options.plus or RFC3986.plus
 
-    local pattern = unreserved(options.unreserved)
-    local unreserved = {}
-    for i = 0, 255 do
-        local ch = string.char(i)
-        unreserved[ch] = ch:match(pattern)
-    end
-
-    return (buf:gsub("(.)", function(ch)
-        if options.plus and ch == " " then
-            return "+"
-        elseif unreserved[ch] then
-            return ch
-        else
-            return char_to_hex(ch)
-        end
-    end))
-end
+    local c_str = ffi.new("char[?]", #buf + 1)
+    ffi.copy(c_str, buf, #buf)
+    local c_escaped = builtin.uri_escape(c_str, #buf, options.unreserved,
+                                         options.plus)
 
-local hex_to_char = function(x)
-    return string.char(tonumber(x, 16))
+    return ffi.string(c_escaped)
```

Comment:
Technically speaking a caller may add a zero byte into the `unreserved` set and a couple of zero bytes may be in the middle of the source string. In this case the result will be truncated.

Of course, unreserved zero byte is strange, but since we anyway need to work with buffer+size in `unescape` (IOW, can't work with zero terminated strings), I would do the same here.